File: flutter/android/app/src/main/kotlin/com/callsafe/mobile/channels/WebRTCChannelHandler.kt
package com.callsafe.mobile.channels

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import org.webrtc.IceCandidate
import org.webrtc.MediaStream
import org.webrtc.SessionDescription
import com.callsafe.mobile.managers.WebRTCManager

/**
 * WebRTC platform channel handler
 * Phase 2: Integrates actual WebRTC implementation
 */
class WebRTCChannelHandler(
    messenger: BinaryMessenger,
    private val context: Context
) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(messenger, CHANNEL_NAME)
    private val eventChannel = EventChannel(messenger, EVENT_CHANNEL_NAME)
    private var eventSink: EventChannel.EventSink? = null

    private var webrtcManager: WebRTCManager? = null

    init {
        channel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initializePeerConnection" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                initializePeerConnection(callAttemptId, result)
            }
            "createAnswer" -> {
                val callAttemptId = call.argument<String>("callAttemptId") ?: return result.error("INVALID_ARGS", "Missing callAttemptId", null)
                val sdpMap = call.argument<Map<String, Any>>("offer") ?: return result.error("INVALID_ARGS", "Missing offer", null)
                val sdpType = sdpMap["type"] as? String ?: return result.error("INVALID_ARGS", "Missing SDP type", null)
                val sdp = sdpMap["sdp"] as? String ?: return result.error("INVALID_ARGS", "Missing SDP", null)

                val offer = SessionDescription(
                    SessionDescription.Type.fromCanonicalForm(sdpType),
                    sdp
                )
                webrtcManager?.createAnswer(offer, callAttemptId)
                result.success(null)
            }
            "addIceCandidate" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                val candidateMap = call.argument<Map<String, Any>>("candidate") ?: return result.error("INVALID_ARGS", "Missing candidate", null)
                val candidateSdp = candidateMap["candidate"] as? String ?: return result.error("INVALID_ARGS", "Missing candidate SDP", null)
                val sdpMLineIndex = (candidateMap["sdpMLineIndex"] as? Number)?.toInt() ?: return result.error("INVALID_ARGS", "Missing sdpMLineIndex", null)
                val sdpMid = candidateMap["sdpMid"] as? String ?: return result.error("INVALID_ARGS", "Missing sdpMid", null)

                val iceCandidate = IceCandidate(sdpMid, sdpMLineIndex, candidateSdp)
                webrtcManager?.addIceCandidate(iceCandidate)
                result.success(null)
            }
            "closePeerConnection" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                webrtcManager?.cleanup()
                webrtcManager = null
                result.success(null)
            }
            "setAudioEnabled" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                val enabled = call.argument<Boolean>("enabled") ?: false
                webrtcManager?.setMicrophoneEnabled(enabled)
                result.success(null)
            }
            "setSpeakerEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                webrtcManager?.setSpeakerEnabled(enabled)
                result.success(null)
            }
            "getMediaCapabilities" -> {
                result.success(mapOf(
                    "canSendAudio" to true,
                    "canSendVideo" to false, // Audio only for now
                    "canReceiveAudio" to true,
                    "canReceiveVideo" to false
                ))
            }
            else -> result.notImplemented()
        }
    }

    private fun initializePeerConnection(callAttemptId: String?, result: MethodChannel.Result) {
        webrtcManager = WebRTCManager(context)

        val webrtcListener = object : WebRTCManager.WebRTCListener {
            override fun onConnectionEstablished() {
                eventSink?.success(mapOf(
                    "type" to "connectionEstablished",
                    "callAttemptId" to callAttemptId
                ))
            }

            override fun onConnectionFailed(error: String) {
                eventSink?.success(mapOf(
                    "type" to "connectionFailed",
                    "callAttemptId" to callAttemptId,
                    "error" to error
                ))
            }

            override fun onRemoteStreamReceived(stream: MediaStream) {
                eventSink?.success(mapOf(
                    "type" to "remoteStreamReceived",
                    "callAttemptId" to callAttemptId
                ))
            }
        }

        val signalingListener = object : WebRTCManager.SignalingListener {
            override fun onAnswer(callAttemptId: String, sdpType: String, sdp: String) {
                eventSink?.success(mapOf(
                    "type" to "answer",
                    "callAttemptId" to callAttemptId,
                    "sdpType" to sdpType,
                    "sdp" to sdp
                ))
            }

            override fun onIceCandidate(callAttemptId: String, candidate: String, sdpMLineIndex: Int, sdpMid: String) {
                eventSink?.success(mapOf(
                    "type" to "iceCandidate",
                    "callAttemptId" to callAttemptId,
                    "candidate" to candidate,
                    "sdpMLineIndex" to sdpMLineIndex,
                    "sdpMid" to sdpMid
                ))
            }
        }

        webrtcManager?.initialize(webrtcListener, signalingListener)
        result.success(null)
    }

    fun dispose() {
        webrtcManager?.cleanup()
        webrtcManager = null
        channel.setMethodCallHandler(null)
    }

    companion object {
        private const val CHANNEL_NAME = "com.callsafe.webrtc"
        private const val EVENT_CHANNEL_NAME = "com.callsafe.webrtc.events"
    }
}

File: flutter/ios/Runner/Channels/WebRTCChannelHandler.swift
import Foundation
import Flutter
import WebRTC

class WebRTCChannelHandler: NSObject, FlutterStreamHandler {
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    private var webrtcManager: WebRTCManager?

    init(messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.callsafe.webrtc",
            binaryMessenger: messenger
        )
        eventChannel = FlutterEventChannel(
            name: "com.callsafe.webrtc.events",
            binaryMessenger: messenger
        )
        super.init()

        methodChannel.setMethodCallHandler(handle)
        eventChannel.setStreamHandler(self)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initializePeerConnection":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing callAttemptId", details: nil))
                return
            }
            initializePeerConnection(callAttemptId: callAttemptId, result: result)

        case "createAnswer":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String,
                  let offerDict = args["offer"] as? [String: Any],
                  let sdpType = offerDict["type"] as? String,
                  let sdp = offerDict["sdp"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }

            let offer = RTCSessionDescription(type: .offer, sdp: sdp)
            webrtcManager?.createAnswer(offer: offer, callAttemptId: callAttemptId)
            result(nil)

        case "addIceCandidate":
            guard let args = call.arguments as? [String: Any],
                  let candidateDict = args["candidate"] as? [String: Any],
                  let candidateSdp = candidateDict["candidate"] as? String,
                  let sdpMLineIndex = candidateDict["sdpMLineIndex"] as? Int,
                  let sdpMid = candidateDict["sdpMid"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid candidate", details: nil))
                return
            }

            let candidate = RTCIceCandidate(
                sdp: candidateSdp,
                sdpMLineIndex: Int32(sdpMLineIndex),
                sdpMid: sdpMid
            )
            webrtcManager?.addIceCandidate(candidate)
            result(nil)

        case "closePeerConnection":
            webrtcManager?.cleanup()
            webrtcManager = nil
            result(nil)

        case "setAudioEnabled":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing enabled", details: nil))
                return
            }
            webrtcManager?.setMicrophoneEnabled(enabled)
            result(nil)

        case "setSpeakerEnabled":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing enabled", details: nil))
                return
            }
            webrtcManager?.setSpeakerEnabled(enabled)
            result(nil)

        case "getMediaCapabilities":
            result([
                "canSendAudio": true,
                "canSendVideo": false,
                "canReceiveAudio": true,
                "canReceiveVideo": false
            ])

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initializePeerConnection(callAttemptId: String, result: @escaping FlutterResult) {
        webrtcManager = WebRTCManager()
        webrtcManager?.delegate = self
        webrtcManager?.initializePeerConnection(callAttemptId: callAttemptId)
        result(nil)
    }

    func dispose() {
        webrtcManager?.cleanup()
        webrtcManager = nil
        methodChannel.setMethodCallHandler(nil)
    }

    // MARK: - FlutterStreamHandler
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

// MARK: - WebRTCManagerDelegate
extension WebRTCChannelHandler: WebRTCManagerDelegate {
    func webRTCManager(_ manager: WebRTCManager, didGenerateIceCandidate candidate: RTCIceCandidate, forCallAttemptId callAttemptId: String) {
        eventSink?([
            "type": "iceCandidate",
            "callAttemptId": callAttemptId,
            "candidate": candidate.sdp,
            "sdpMLineIndex": candidate.sdpMLineIndex,
            "sdpMid": candidate.sdpMid ?? ""
        ])
    }

    func webRTCManager(_ manager: WebRTCManager, didCreateAnswer sdp: RTCSessionDescription, forCallAttemptId callAttemptId: String) {
        eventSink?([
            "type": "answer",
            "callAttemptId": callAttemptId,
            "sdpType": sdp.type.string(),
            "sdp": sdp.sdp
        ])
    }

    func webRTCManager(_ manager: WebRTCManager, didChangeConnectionState state: RTCIceConnectionState) {
        switch state {
        case .connected, .completed:
            eventSink?([
                "type": "connectionEstablished"
            ])
        default:
            break
        }
    }

    func webRTCManager(_ manager: WebRTCManager, didReceiveRemoteStream stream: RTCMediaStream) {
        eventSink?([
            "type": "remoteStreamReceived"
        ])
    }

    func webRTCManager(_ manager: WebRTCManager, didFailWithError error: String) {
        eventSink?([
            "type": "connectionFailed",
            "error": error
        ])
    }
}

// MARK: - RTCSdpType Extension
extension RTCSdpType {
    func string() -> String {
        switch self {
        case .offer:
            return "offer"
        case .prAnswer:
            return "pranswer"
        case .answer:
            return "answer"
        case .rollback:
            return "rollback"
        @unknown default:
            return "unknown"
        }
    }
}
